### PR TITLE
#2771 + #3015: Create requesting entity page - [MEOWARD]

### LINKED ISSUE #2771: (DRAFT) Create Requesting Entity Page
### Issue description

Create the Requesting entity page in its entirety, following [the figma specs](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=7631-26542&node-type=frame&t=yQTNZD3MVr5e8bJ6-0).
Also add “Requesting Entity” section to “Review and Submit” page and to e-mail.

### Acceptance criteria

- [ ] Create the Requesting entity page so that it is the first step in a Portfolio Domain Request form
- [ ] For a *brand new* request form, both radio buttons are unselected by default
- [ ] Selecting a radio button updates the page according to the figma annotations
- [ ] The Requesting entity step in the side nav locks and unlocks just like the other steps in our regular Domain Request flow (Editing an in-progress Portfolio Domain Request that already had Requesting Entity data will load the request form with the Requesting Entity step unlocked.)
- [ ] Required fields are flagged as required and follow the same default error handling message format as other required fields in our regular Domain Request flow

### Additional context

_No response_

### Links to other issues

- ⛔️ Is blocked by #2760

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #2771 and #3015

## Changes

- Added a "requesting entity" domain request form step + added page to pa11y
- Added new fields to domainrequest and domain information: requested_suborganization, suborganization_city, suborganization_state_territory
- Added new fields to django admin, and only show them for superusers (for now)
- Added javascript to show/hide the new fields in django admin when the user is making a new suborganization request
- Added javascript for special form display logic on the requesting entity page
- Added a custom submission email for organization domain requests (toggles between new and old if a portfolio exists)
- Updated the request review page to include data from the requesting entity step (with logic for if a portfolio is selected vs a suborg vs a custom one)
- Added the new review page to the portfolio status manage page (as it was using the old non-portfolio one)
- Updated the domain request status page to display the new summary

## Context for reviewers
Primarily, this PR does four things for portfolio domain requests:
1. Adds a new requesting entity step
2. Updates the existing request summaries to include this step
3. Adds a new submission confirmation email to instead include information from the portfolio request
4. Updates the domain request status page (and viewonly page) summaries to match the new request (#3015)
5. Adds the new requesting_suborganization, suborganization_city, and suborganization_state_territory fields to django admin. They are readonly for analysts, and these fields are hidden when the organization flag is off, when no portfolio is selected, or when a portfolio is selected and a suborganization is selected

Note: this PR does not include logic to automatically create a suborganization when a domain request is approved. This is intentional as there currently is not a determination on if that should be done.

## Setup
**Testing requesting entity**
First, enable the organization_feature and organization_request flags. Then start a new domain request. Requesting entity should be the first option you see.
There are three flows on this page. These all should match [the figma](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=7631-26542&node-type=frame&t=yQTNZD3MVr5e8bJ6-0).

**Testing the domain request summaries**
After testing the step above, complete the domain request and go to the review your domain request page. These should match the figma. The only exception to note is that if you just select a suborganization, no state or city will display as there is no data to pull from (another ticket will add that to the suborganization object).

Additionally, verify that this summary is displayed correctly on the "view" and "manage" contexts. You can test the manage context by submitting your domain, and the view by just clicking on a random domain request on the portfolio. (#3015)

**Testing Django Admin**
The requesting_entity, requesting_suborganization, and requesting_state_territory fields should only be visible when the domain request has a portfolio and when no suborganization is selected. Additionally, these fields should only be visible for superusers as of now. 

**Testing email**
Submit a domain request and ensure that all data displays correctly. You should see a new email with new content.

- Ensure that these new fields are hidden on django admin with the org flag off

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Update documentation in READMEs and/or onboarding guide

#### Ensured code standards are met (Original Developer)

- [x] If any updated dependencies on Pipfile, also update dependencies in requirements.txt.
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [x] Tag @dotgov-designers in this PR's Reviewers for design review. If code is not user-facing, delete design reviewer checklist
- [x] Verify new pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Verified code meets all checks above. Address any checks that are not satisfied
- [ ] Reviewed this code and left comments. Indicate if comments must be addressed before code is merged
- [ ] Checked that all code is adequately covered by tests
- [ ] Verify migrations are valid and do not conflict with existing migrations

#### Validated user-facing changes as a developer
**Note:** Multiple code reviewers can share the checklists above, a second reviewer should not make a duplicate checklist. All checks should be checked before approving, even those labeled N/A. 

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] (Rarely needed) Tested as both an analyst and applicant user

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior. Comment any found issues or broken flows.
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### References
- [Code review best practices](../docs/dev-practices/code_review.md)

### CODE REVIEW

File: src/registrar/templates/includes/portfolio_status_manage.html [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{% extends "includes/request_status_manage.html" %}
```
Comment (rachidatecs) [inline src/registrar/templates/includes/portfolio_status_manage.html:1]: Why do we need an interstitial that only passes the baton to another include? Can't this block live in domain_request_status.html directly?

File: src/registrar/views/domain_request.py [hunk 2]
```diff
@@ -834,6 +862,18 @@ def has_permission(self):
 
         return True
 
+    def get_context_data(self, **kwargs):
```
Comment (rachidatecs) [inline src/registrar/views/domain_request.py:865]: Q: Why is this needed?

File: src/registrar/tests/test_views_portfolio.py [hunk 3]
```diff
@@ -1592,3 +1595,285 @@ def test_member_domains_not_found(self):
...
+        # Ensure that the post occurred successfully by checking that we're on the following page.
+        self.assertContains(response, "Current websites")
+        created_domain_request_exists = DomainRequest.objects.filter(
+            organization_name__isnull=True,
+            sub_organization__isnull=True,
+            requested_suborganization="moon",
+            suborganization_city="kepler",
+            suborganization_state_territory=DomainRequest.StateTerritoryChoices.ALABAMA,
+        ).exists()
+        self.assertTrue(created_domain_request_exists)
```
Comment (rachidatecs) [inline src/registrar/tests/test_views_portfolio.py:1742]: Nice!

File: src/registrar/tests/test_views_request.py [hunk 4]
```diff
@@ -2887,8 +2887,6 @@ def test_domain_request_withdraw_portfolio_redirects_correctly(self):
         detail_page = self.app.get(f"/domain-request/{domain_request.id}")
         self.assertContains(detail_page, "city.gov")
         self.assertContains(detail_page, "city1.gov")
-        self.assertContains(detail_page, "Chief Tester")
```
Comment (rachidatecs) [inline src/registrar/tests/test_views_request.py:2890]: Q: What broke these?

File: src/registrar/models/domain_information.py [hunk 5]
```diff
@@ -75,6 +75,24 @@ class Meta:
         verbose_name="Suborganization",
     )
 
+    requested_suborganization = models.CharField(
```
Comment (abroddrick) [inline src/registrar/models/domain_information.py:78]: **blocking** please remove this from domain info, we don't need to store it here.

Comment (abroddrick) [reply src/registrar/models/domain_information.py:78]: this is per [this](https://cisa-corp.slack.com/archives/C05BGB4L5NF/p1730392283781909) slack thread

File: src/registrar/forms/domain_request_wizard.py [hunk 6]
```diff
@@ -22,10 +22,151 @@
...
+
+    # IMPORTANT: This is tied to DomainRequest.is_requesting_new_suborganization().
+    # This is due to the from_database method on DomainRequestWizard.
+    # Add a hidden field to store if the user is requesting a new suborganization.
+    # This hidden boolean is used for our javascript to communicate to us and to it.
+    # If true, the suborganization form will auto select a js value "Other".
+    # If this selection is made on the form (tracked by js), then it will toggle the form value of this.
+    # In other words, this essentially tracks if the suborganization field == "Other".
+    # "Other" is just an imaginary value that is otherwise invalid.
+    # Note the logic in `def clean` and line 2744 in get-gov.js
```
Comment (abroddrick) [inline src/registrar/forms/domain_request_wizard.py:38]: **comment**: don't include line numbers in code comments please, they will fall out of date

File: src/registrar/forms/domain_request_wizard.py [hunk 7]
```diff
@@ -22,10 +22,151 @@
...
+    is_requesting_new_suborganization = forms.BooleanField(required=False, widget=forms.HiddenInput())
+
+    sub_organization = forms.ModelChoiceField(
+        label="Suborganization name",
+        required=False,
+        queryset=Suborganization.objects.none(),
+        empty_label="--Select--",
```
Comment (abroddrick) [inline src/registrar/forms/domain_request_wizard.py:45]: **change before merging:** add an error message here see [response](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=7631-26542&node-type=frame&t=mnJyTGfVpTMoWGba-0) to you question here

File: src/registrar/forms/domain_request_wizard.py [hunk 8]
```diff
@@ -22,10 +22,151 @@
...
+        )
+        if requesting_entity_is_suborganization == "True":
+            if is_requesting_new_suborganization:
+                # Validate custom suborganization fields
+                if not cleaned_data.get("requested_suborganization"):
+                    self.add_error("requested_suborganization", "Enter details for your organization name.")
+                if not cleaned_data.get("suborganization_city"):
+                    self.add_error("suborganization_city", "Enter details for your city.")
+                if not cleaned_data.get("suborganization_state_territory"):
+                    self.add_error("suborganization_state_territory", "Enter details for your state or territory.")
```
Comment (abroddrick) [inline src/registrar/forms/domain_request_wizard.py:127]: **blocking** Why don't these errors match the ones you have above in the field definitions? If the others aren't triggering due to the setup, they should be removed to avoid confusion and the content copied here.

File: src/registrar/forms/domain_request_wizard.py [hunk 9]
```diff
@@ -22,10 +22,151 @@
...
+
+    @property
+    def form_is_checked(self):
+        """
+        Determines if the requesting entity is a suborganization, or a portfolio.
+        For suborganizations, users have the ability to request a new one if the
+        desired suborg doesn't exist. We expose additional fields that denote this,
+        like `requested_suborganization`. So we also check on those.
+        """
+        # True means that the requesting entity is a suborganization,
```
Comment (abroddrick) [inline src/registrar/forms/domain_request_wizard.py:162]: **suggestion** Key info on your return values should exist in the docstring for the function. 
example to shorten it here
"""
Determines the initial checked state of the form, returning True (checked) if the requesting entity is a suborganization,  and False if it is a portfolio. Returns None if neither condition is met.

"""

**Comment by abroddrick** [comment]
@zandercymatics you have a period in your radio field that is not in the figma and it makes the capitalization look off. Just remove the period please. 
yours:
image
the figma: 
image

File: src/registrar/admin.py [hunk 10]
```diff
@@ -1868,6 +1885,28 @@ def status_history(self, obj):
 
     change_form_template = "django/admin/domain_request_change_form.html"
 
+    # While the organization feature is under development, we can gate some fields
+    # from analysts for now. Remove this array and the get_fieldset overrides once this is done.
+    # Not my code initially, credit to Nicolle. This was once removed and like a phoenix it has been reborn.
+    superuser_only_fields = [
```
Comment (rachidatecs) [inline src/registrar/admin.py:1891]: I did not catch this on the first pass but I think we should simply have these fields as readonly for analysts. This AC indicates so:

_The Admin Readonly view for Portfolio Request, Domain info/Domain does not show the requested suborganization, suborg_city and state fields if the user is selecting a suborganization from the list. But it displays the requested suborganization, city/state fields when the user selects the "Other" option._

If that's the case, we should use pre-established patters for analyst readonly / superuser editable fields.

File: src/registrar/views/domain_request.py (hunk 2 above)
Comment (rachidatecs) [reply src/registrar/views/domain_request.py:873]: This is too out of context for me to review, so I will withhold opinion on this code. If you NEED it reviewed, let's catch up on it tomorrow otherwise I will ignore it in my review (won't block)

**Review by rachidatecs — COMMENTED** [review]
Possibly/probably out of scope: The admin does not logically reflect business logic.

Scenario 1: Suborganization should not show as blank. Maybe 'Other'?

Screenshot 2024-11-01 at 1 22 10 AM

Screenshot 2024-11-01 at 1 22 26 AM

Scenario 2: I am able to select a suborg that does not belong to the portfolio
Screenshot 2024-11-01 at 1 33 37 AM

Issue 3:
Nav on the request page is missing all links except for Members?
Screenshot 2024-11-01 at 1 35 30 AM

Issue 4: Sidenav seems buggy on Purpose and Additional details on the org request page:
Screenshot 2024-11-01 at 1 36 09 AM

Defer to @abroddrick  on whether these are included in future tickets / are bugs / in scope.

File: src/registrar/templates/includes/portfolio_request_review_steps.html [hunk 11]
```diff
@@ -8,15 +8,24 @@
     {% endif %}
 
     {% if step == Step.REQUESTING_ENTITY %}
-        {% if domain_request.organization_name %}
-            {% with title=form_titles|get_item:step value=domain_request %}
-                {% include "includes/summary_item.html" with title=title value=value heading_level=heading_level editable=is_editable edit_link=domain_request_url address='true' %}
-            {% endwith %}
+        {% with title=form_titles|get_item:step %}
+        {% if domain_request.sub_organization %}
+            {% include "includes/summary_item.html" with value=domain_request.sub_organization edit_link=domain_request_url %}
```
Comment (rachidatecs) [inline src/registrar/templates/includes/portfolio_request_review_steps.html:13]: Q:
Selecting top level organization shows its city/state on the summary page. Same for selection a other suborg, it shows the custom suborg city state. Selecting an existing suborg is the only scenario that does not show a city/state. 
 
Screenshot 2024-11-01 at 1 43 41 AM

File: src/registrar/admin.py (hunk 10 above)
Comment (abroddrick) [reply src/registrar/admin.py]: I would agree read only better follows our pre-existing patterns. However, I do think design wants those fields to not show. That said, @Katherine-Osos could we have the requested suborganization, suborg_city and state fields just turn read-only when suborganization is filled in and still provide a similiar user experience for analysts?

File: src/registrar/templates/includes/portfolio_request_review_steps.html (hunk 11 above)
Comment (abroddrick) [reply src/registrar/templates/includes/portfolio_request_review_steps.html:13]: City and state definitely shouldn't show if the user selected another (real) suborg.

**Comment by abroddrick** [comment]
Responding to @rachidatecs 's message

> Scenario 1: Suborganization should not show as blank. Maybe 'Other'?

In django admin, we do want it blank so that an analysts can easily see that suborg needs to be selected. We also don't want to add in an "other" record into the suborganization table, so inorder to show "other" on django admin we would need to do more js to customize admin to show other. This doesn't seem needed and just increases the js complexity on admin. 

> Scenario 2: I am able to select a suborg that does not belong to the portfolio

Funny enough this is an AC on [the ticket](https://github.com/cisagov/manage.get.gov/issues/2934) you (Rachid) and Dave are assigned to "the "suborganization" field should only show suborgs associated with the selected portfolio."

> Issue 3:
> Nav on the request page is missing all links except for Members?
That's a good catch! The figma shows that nothing should show in that top nav besides organization (which i disagree with a bit, given it makes navigating back really unintuitive). @zandercymatics I believe this was actually done in your ticket for the first domain request ticket. Did maybe some bug occur making the top nav incorrect or was that AC always missed?
the figma: 
image

4. Sidenav seems buggy on Purpose and Additional details on the org request page:
Did you have those fields filled in at all, or were you on a brand new request? I would suggest @zandercymatics double check that this does exist on other sandboxes. If this is introduced in this PR we want to make sure to fix it.

File: src/registrar/migrations/0136_domaininformation_requested_suborganization_and_more.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
+# Generated by Django 4.2.10 on 2024-10-30 17:59
```
Comment (rachidatecs) [inline src/registrar/migrations/0136_domaininformation_requested_suborganization_and_more.py:1]: BLOCKING: Delete both migrations and rerun makemigrations to generate 1 file

File: src/registrar/forms/domain_request_wizard.py (hunk 8 above)
Comment (rachidatecs) [reply src/registrar/forms/domain_request_wizard.py:122]: Q: Why aren't we using error definitions on the form fields instead?

File: src/registrar/templates/includes/portfolio_request_review_steps.html [hunk 13]
```diff
@@ -8,15 +8,24 @@
...
+            {% comment %} We don't have city or state_territory for suborganizations yet, so no data should display {% endcomment %}
+        {% elif domain_request.requesting_entity_is_suborganization %}
+            {% include "includes/summary_item.html" with value=domain_request.requested_suborganization edit_link=domain_request_url %}
+            <p class="margin-y-0">{{domain_request.suborganization_city}}, {{domain_request.suborganization_state_territory}}</p>
+        {% elif domain_request.requesting_entity_is_portfolio %}
+            {% include "includes/summary_item.html" with value=domain_request.portfolio.organization_name edit_link=domain_request_url %}
+            {% if domain_request.portfolio.city and domain_request.portfolio.state_territory %}
+                <p class="margin-top-0 margin-bottom-0">{{domain_request.portfolio.city}}, {{domain_request.portfolio.state_territory}}</p>
```
Comment (rachidatecs) [inline src/registrar/templates/includes/portfolio_request_review_steps.html:21]:
```suggestion
                <p class="margin-y-0">{{domain_request.portfolio.city}}, {{domain_request.portfolio.state_territory}}</p>
```
